internal/controllers/machinedeployment/machinedeployment_inplace.go

```diff
@@ -3,42 +3,66 @@ package machinedeployment
 import (
 	"context"
 
-	"github.com/pkg/errors"
-	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/internal/controllers/machinedeployment/mdutil"
 	"sigs.k8s.io/cluster-api/util/annotations"
 )
 
 func (r *Reconciler) rolloutInPlace(ctx context.Context, md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (reterr error) {
 	log := ctrl.LoggerFrom(ctx)
 
-	// For in-place upgrade, we shouldn't try to create a new MachineSet as that would trigger a rollout.
-	// Instead, we should try to get latest MachineSet that matches the MachineDeployment.Spec.Template/
+	// If there are no MachineSets for a MachineDeployment, either this is a create operation for a new
+	// MachineDeployment or the MachineSets were manually deleted. In either case, a new MachineSet should be created
+	// as there are no MachineSets that can be in-place upgraded.
+	// If there are already MachineSets present, we shouldn't try to create a new MachineSet as that would trigger a rollout.
+	// Instead, we should try to get latest MachineSet that matches the MachineDeployment.Spec.Template
 	// If no such MachineSet exists yet, this means the MachineSet hasn't been in-place upgraded yet.
 	// The external in-place upgrade implementer is responsible for updating the latest MachineSet's template
 	// after in-place upgrade of all worker nodes belonging to the MD is complete.
 	// Once the MachineSet is updated, this function will return the latest MachineSet that matches the
 	// MachineDeployment template and thus we can deduce that the in-place upgrade is complete.
-	newMachineSet, oldMachineSets, err := r.getAllMachineSetsAndSyncRevision(ctx, md, msList, false)
+	newMachineSetNeeded := len(msList) == 0
+	newMachineSet, oldMachineSets, err := r.getAllMachineSetsAndSyncRevision(ctx, md, msList, newMachineSetNeeded)
 	if err != nil {
 		return err
 	}
 
-	defer func() {
-		allMSs := append(oldMachineSets, newMachineSet)
-
-		// Always attempt to sync the status
-		err := r.syncDeploymentStatus(allMSs, newMachineSet, md)
-		reterr = kerrors.NewAggregate([]error{reterr, err})
-	}()
+	allMSs := append(oldMachineSets, newMachineSet)
```

Comment:
Can newMachineSet be nil here. In that case should we append it here.?

Reply:
CAPI handles that MachineSet be nil since they are pointers so it's fine if it's nil https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/mdutil/util.go#L474

Reply:
Yea but why not add to the if condition below then? We could then only set allMSs to oldMachineSets and append in that if condition that's already doing a nil check?

Reply:
Also in the same method you linked below, they also don't add if it's nil even though we have a check https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/machinedeployment_rolling.go#L41-L43

Reply:
okay, yeah we can do that. Updated